src/runtime_src/core/edge/user/aie/graph.cpp

```diff
@@ -261,34 +266,37 @@ update_rtp(const char* port, const char* buffer, size_t size)
     auto rtp = std::find_if(rtps.begin(), rtps.end(),
             [port](rtp_type it) { return it.name.compare(port) == 0; });
 
-    if (rtp != rtps.end())
+    if (rtp == rtps.end())
       throw xrt_core::error(-EINVAL, "Can't update graph '" + name + "': RTP port '" + port + "' not found");
 
     if (rtp->is_plrtp)
       throw xrt_core::error(-EINVAL, "Can't update graph '" + name + "': not aie rtp port");
- 
-    /* We only support input RTP */
+
     if (!rtp->is_input)
-      throw xrt_core::error(-EINVAL, "Can't update graph '" + name + "': only input RTP port supported");
- 
-    /*
-     * Only support sync update
-     * TODO support async update
-     */
-    if (rtp->is_async)
-      throw xrt_core::error(-EINVAL, "Can't update graph '" + name + "': only synchronous update supported");
- 
+      throw xrt_core::error(-EINVAL, "Can't update graph '" + name + "': RTP port '" + port + "' is not input");
+
     /* If RTP port is connected, only support async update */
-    if (rtp->is_connected)
-      throw xrt_core::error(-EINVAL, "Can't update graph '" + name + "': connected sync update is not supported");
- 
+    if (rtp->is_connected) {
+        if (rtp->is_async && state == graph_state::running)
+            throw xrt_core::error(-EINVAL, "Can't update graph '" + name + "': connected async update is not supported while graph is running");
+        if (!rtp->is_async)
+            throw xrt_core::error(-EINVAL, "Can't update graph '" + name + "': connected sync update is not supported");
+    }
+
     auto selector_pos = aieArray->getTilePos(rtp->selector_col, rtp->selector_row);
     auto selector_tile = &(aieArray->tileArray.at(selector_pos));
 
-    if (rtp->require_lock)
+    /* Don't acquire selector lock for async RTP while graph is not running */
+    bool need_lock = rtp->require_lock && (
+        rtp->is_async && state == graph_state::running ||
+        !rtp->is_async
+	);
+
+    if (need_lock)
         XAieTile_LockAcquire(selector_tile, rtp->selector_lock_id, 0, 0xFFFFFFFF);
```

Comment:
LockAcquire and LockRelease is not a good pattern.  What happens in an exception is thrown in between?
Don't worry about it now, we can fix this later.

Reply:
Yeah. I agree with you. Hopefully a later AIE driver could hide the lock acquire/release from the interface (better in kernel space).